File: frontend/src/stores/fantasy.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { useApi } from "@/composables/useApi";

export interface FantasyPick {
  id: string;
  team: string;
  match_id: string;
  matchday_number: number;
  goals_scored: number | null;
  goals_conceded: number | null;
  match_result: string | null;
  fantasy_points: number | null;
  status: string;
}

export interface FantasyStanding {
  user_id: string;
  alias: string;
  total_points: number;
  matchdays_played: number;
  avg_points: number;
}

export const useFantasyStore = defineStore("fantasy", () => {
  const api = useApi();

  const pick = ref<FantasyPick | null>(null);
  const standings = ref<FantasyStanding[]>([]);
  const loading = ref(false);

  async function fetchPick(
    squadId: string,
    sport: string,
    season: number,
    matchdayNumber: number,
  ) {
    loading.value = true;
    try {
      pick.value = await api.get<FantasyPick | null>(`/fantasy/${squadId}/pick`, {
        sport,
        season: String(season),
        matchday_number: String(matchdayNumber),
      });
    } catch {
      pick.value = null;
    } finally {
      loading.value = false;
    }
  }

  async function makePick(
    squadId: string,
    matchId: string,
    team: string,
  ): Promise<FantasyPick> {
    const result = await api.post<FantasyPick>(`/fantasy/${squadId}/pick`, {
      match_id: matchId,
      team,
    });
    pick.value = result;
    return result;
  }

  async function fetchStandings(squadId: string, sport: string, season?: number) {
    try {
      const params: Record<string, string> = { sport };
      if (season) params.season = String(season);
      standings.value = await api.get<FantasyStanding[]>(
        `/fantasy/${squadId}/standings`,
        params,
      );
    } catch {
      standings.value = [];
    }
  }

  function reset() {
    pick.value = null;
    standings.value = [];
  }

  return {
    pick,
    standings,
    loading,
    fetchPick,
    makePick,
    fetchStandings,
    reset,
  };
});

File: backend/app/routers/fantasy.py
"""Fantasy Matchups endpoints — pick teams, view standings."""

from fastapi import APIRouter, Depends, Query, status

from app.models.wallet import FantasyPickCreate, FantasyPickResponse
from app.services import fantasy_service
from app.services.auth_service import get_current_user
from app.utils import utcnow

router = APIRouter(prefix="/api/fantasy", tags=["fantasy"])


@router.post("/{squad_id}/pick", status_code=status.HTTP_201_CREATED)
async def make_pick(
    squad_id: str,
    body: FantasyPickCreate,
    user=Depends(get_current_user),
):
    """Make a fantasy team pick for the current matchday."""
    user_id = str(user["_id"])
    pick = await fantasy_service.make_pick(
        user_id=user_id,
        squad_id=squad_id,
        match_id=body.match_id,
        team=body.team,
    )
    return FantasyPickResponse(
        id=str(pick["_id"]),
        team=pick["team"],
        match_id=pick["match_id"],
        goals_scored=pick.get("goals_scored"),
        goals_conceded=pick.get("goals_conceded"),
        fantasy_points=pick.get("fantasy_points"),
        status=pick["status"],
        created_at=pick["created_at"],
    )


@router.get("/{squad_id}/pick")
async def get_pick(
    squad_id: str,
    sport: str = Query(...),
    season: int = Query(None),
    matchday: int = Query(...),
    user=Depends(get_current_user),
):
    """Get user's fantasy pick for a specific matchday."""
    user_id = str(user["_id"])
    if not season:
        season = utcnow().year

    pick = await fantasy_service.get_user_pick(user_id, squad_id, sport, season, matchday)
    if not pick:
        return None
    return FantasyPickResponse(
        id=str(pick["_id"]),
        team=pick["team"],
        match_id=pick["match_id"],
        goals_scored=pick.get("goals_scored"),
        goals_conceded=pick.get("goals_conceded"),
        fantasy_points=pick.get("fantasy_points"),
        status=pick["status"],
        created_at=pick["created_at"],
    )


@router.get("/{squad_id}/standings")
async def get_standings(
    squad_id: str,
    sport: str = Query(...),
    season: int = Query(None),
    user=Depends(get_current_user),
):
    """Get fantasy standings for a squad."""
    if not season:
        season = utcnow().year
    return await fantasy_service.get_standings(squad_id, sport, season)
